File: Go/Day17.go
package main

import AH "./adventhelper"

type Pos struct {
	X, Y, Z, T int
}

func (pt Pos) nbrs(dimn int) (ns []Pos) {
	ns = []Pos{}

	for x := -1; x < 2; x++ {
		for y := -1; y < 2; y++ {
			for z := -1; z < 2; z++ {
				if dimn == 3 {
					if x == 0 && y == 0 && z == 0 {
						continue
					}
					ns = append(ns, Pos{X:pt.X + x, Y:pt.Y + y, Z:pt.Z + z})
				}
				if dimn == 4 {
					for t := -1; t < 2; t++ {
						if x == 0 && y == 0 && z == 0 && t == 0 {
							continue
						}
						ns = append(ns, Pos{X:pt.X + x, Y:pt.Y + y,
						                    Z:pt.Z + z, T:pt.T + t})
					}
				}
			}
		}
	}

	return
}

func (pt Pos) nbrsCount(space map[Pos]int, dimn int) (count int) {
	count = 0
	ns := pt.nbrs(dimn)
	for _, n := range ns {
		if _, ok := space[n]; ok {
			count++
		}
	}
	return
}

func BuildNSpace (ss []string, dimn int) (space map[Pos]int) {
	space = make(map[Pos]int)
	for i, s := range ss {
		for j, c := range s {
			if c == '#' {
				if dimn	 == 3 {
					space[Pos{X:i, Y:j, Z:0}] = 1
				} else if dimn == 4 {
					space[Pos{X:i, Y:j, Z:0, T:0}] = 1
				}
			}
		}
	}
	return
}

func tick (space map[Pos]int, dimn int) (newSpace map[Pos]int) {
	newSpace = make(map[Pos]int)

	for pos, _ := range space {
		// check if this cell surives
		nCount := pos.nbrsCount(space, dimn) // the number of neighbours
		if nCount == 2 || nCount == 3 {
			newSpace[pos] = 1
		}
		// now check its neighbours
		neighbours := pos.nbrs(dimn)
		for _, nPos := range neighbours {
			nCount = nPos.nbrsCount(space, dimn) // the number of neighbours
			// if we've already done this point...
			if _, ok := newSpace[nPos]; ok {
				// do nothing..
			} else { // we haven't checked this point yet
				if _, ok := space[nPos]; ok { // Is it currently alive..
					if nCount == 2 || nCount == 3 { //..and has 2 or 3 live neighbours...
						newSpace[nPos] = 1 // ..send it to next generation
					} // otherwise it dies here
				} else { // It is NOT currently alive...
					if nCount == 3 { //..and has exactly 3 live neighbours...
						newSpace[nPos] = 1 // ..send it to next generation
					} // otherwise it dies here
				}
			}
		}
	}

	return
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input17.txt")
	space3 := BuildNSpace (ss, 3)
	space4 := BuildNSpace (ss, 4)

	for i := 0; i < 6; i++ {
		space3 = tick(space3, 3)
		space4 = tick(space4, 4)
	}

	AH.PrintSoln(17, len(space3), len(space4))

	return
}

File: C++/Day17.cpp
#include "AH.h"

namespace Day17
{

	typedef std::tuple<int64_t, int64_t, int64_t, int64_t> Pos;

	std::vector<Pos> nbrs(const Pos pt, const int64_t dimn)
	{
		std::vector<Pos> ns;
		ns.reserve(1 << dimn);
		for (int64_t x = -1; x < 2; ++x)
		{
			for (int64_t y = -1; y < 2; ++y)
			{
				for (int64_t z = -1; z < 2; ++z)
				{
					if (dimn == 3)
					{
						if (x == 0 && y == 0 && z == 0)
							continue;

						ns.push_back(std::make_tuple(std::get<0>(pt) + x,
						                             std::get<1>(pt) + y,
						                             std::get<2>(pt) + z, 0));
					}
					if (dimn == 4)
					{
						for (int64_t t = -1; t < 2; ++t)
						{
							if (x == 0 && y == 0 && z == 0 && t == 0)
								continue;

							ns.push_back(std::make_tuple(std::get<0>(pt) + x,
							                             std::get<1>(pt) + y,
							                             std::get<2>(pt) + z,
							                             std::get<3>(pt) + t));
						}
					}
				}
			}
		}

		return ns;
	}

	uint64_t nbrsCount(const Pos pt, const std::set<Pos> space, const uint64_t dimn)
	{
		uint64_t count = 0;
		auto ns = nbrs(pt, dimn);
		for (auto n : ns)
		{
			if ( space.find(n) != space.end() )
				count++;
		}
		return count;
	}

	std::set<Pos> BuildNSpace(const std::vector<std::string> ss)
	{
		std::set<Pos> space;
		int64_t i = 0, j = 0;
		for (auto s : ss)
		{
			for (auto c : s)
			{
				if (c == '#')
				{
					auto tpl = std::make_tuple(i, j, 0, 0);
					space.insert(tpl);
				}
				++j;
			}
			++i;
			j = 0;
		}
		return space;
	}

	std::set<Pos> tick(const std::set<Pos> space, const uint64_t dimn)
	{
		std::set<Pos> newSpace;
		// iterate over the live cells
		for (auto pos : space)
		{
			// check if this cell survives
			auto nCountHere = nbrsCount(pos, space, dimn);
			if (nCountHere == 2 || nCountHere == 3)
				newSpace.insert(pos);

			// no check this point's neighbours
			auto neighbours = nbrs(pos, dimn);
			for (auto nPos : neighbours)
			{
				auto nCountThere = nbrsCount(nPos, space, dimn);
				// we've already checked this point...
				if ( newSpace.find(nPos) != newSpace.end() )
				{
					// do nothing
				}
				// we haven't checked this point
				else
				{
					if ( space.find(nPos) != space.end() )
					{
						if (nCountThere == 2 || nCountThere == 3)
							newSpace.insert(nPos);
					}
					else
					{
						if (nCountThere == 3)
							newSpace.insert(nPos);
					}
				}
			}
		}

		return newSpace;
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		auto space3 = BuildNSpace(inputLines);

		auto space4 = BuildNSpace(inputLines);
		for (size_t i = 0; i < 6; ++i)
		{
			space3 = tick(space3, 3);
			space4 = tick(space4, 4);
		}

		AH::PrintSoln(17, space3.size(), space4.size());

		return 0;
	}

}
